### PR TITLE
APEXCORE-592 Return additional parameter description from properties.xml in defaultProperties for get-app-package-info command

### PR DESCRIPTION
**Pull request by ajaygit158-zz**
Sample data and output : 
**properties.xml**
<property>
<name>dt.application.FraudDetectDemo.operator.smaPerMerchant.windowSize
</name>
<value>30</value>
<description>Window size</description>
</property>
<property>
<name>dt.application.FraudDetectDemo.operator.movingSum.attr.APPLICATION_WINDOW_COUNT
</name>
<value>10</value>
<description>Window count for application</description>
</property>

**Output : get-app-package-info package-name**
"dt.application.FraudDetectDemo.operator.smaPerMerchant.windowSize":
{ "value": "30", "description": "Window size" }
,
"dt.application.FraudDetectDemo.operator.movingSum.attr.APPLICATION_WINDOW_COUNT":
{ "value": "10", "description": null }

### CODE REVIEW

**Comment by ajaygit158-zz** [comment]
@vrozov Kindly review.

**Comment by vrozov** [comment]
@sandeshh Please review. @amberarrow See if this affects documentation.

File: engine/src/main/java/com/datatorrent/stram/client/DTConfiguration.java [hunk 1]
```diff
@@ -99,6 +102,11 @@ public ConfigException(String message)
     return result.entrySet().iterator();
   }
 
+  public Map<String, ValueEntry> getMap()
```
Comment (sandeshh) [inline engine/src/main/java/com/datatorrent/stram/client/DTConfiguration.java]: There is no need for getMap()

File: engine/src/main/java/com/datatorrent/stram/client/AppPackage.java [hunk 2]
```diff
@@ -426,21 +427,22 @@ private void processPropertiesXml(File file, AppInfo app)
     DTConfiguration config = new DTConfiguration();
     try {
       config.loadFile(file);
-      for (Map.Entry<String, String> entry : config) {
+      for (Map.Entry<String, ValueEntry> entry : config.getMap().entrySet()) {
         String key = entry.getKey();
-        String value = entry.getValue();
-        if (value == null) {
+        ValueEntry valueEntry = entry.getValue();
+        LOG.info("Properties: {} {} {}", key, valueEntry.value, valueEntry.description);
```
Comment (sandeshh) [inline engine/src/main/java/com/datatorrent/stram/client/AppPackage.java]: LOG.debug?

**Comment by sandeshh** [comment]
Please see my comment in the jira.

File: engine/src/main/java/com/datatorrent/stram/cli/ApexCli.java [hunk 3]
```diff
@@ -3448,13 +3449,25 @@ public void execute(String[] args, ConsoleReader reader) throws Exception
     public void execute(String[] args, ConsoleReader reader) throws Exception
     {
       try (AppPackage ap = newAppPackageInstance(new File(expandFileName(args[1], true)))) {
+        //Ignore all attribute descriptions since attributes are platform dependent and not app-package dependent.
```
Comment (sandeshh) [inline engine/src/main/java/com/datatorrent/stram/cli/ApexCli.java]: Do we even need to filter these values? What if the user wants to have a custom description, as to why they are setting that particular attribute?

Comment (vrozov) [reply engine/src/main/java/com/datatorrent/stram/cli/ApexCli.java]: @sandeshh Please see discussion in JIRA.

File: engine/src/main/java/com/datatorrent/stram/client/AppPackage.java (hunk 2 above)
Comment (vrozov) [reply engine/src/main/java/com/datatorrent/stram/client/AppPackage.java]: Why is the logging necessary, especially for every Map.Entry in the Set?

**Comment by vrozov** [comment]
@ajaygit158 Please address review comments, rebase to upstream/master and squash commits.

File: engine/src/main/java/com/datatorrent/stram/client/AppPackage.java (hunk 2 above)
Comment (ajaygit158-zz) [reply engine/src/main/java/com/datatorrent/stram/client/AppPackage.java]: True. Removed the log

File: engine/src/main/java/com/datatorrent/stram/client/DTConfiguration.java (hunk 1 above)
Comment (ajaygit158-zz) [reply engine/src/main/java/com/datatorrent/stram/client/DTConfiguration.java]: I am using it in AppPackage.java

**Review by ajaygit158-zz — COMMENTED** [review]
Review comments addressed. Code rebased.

**Comment by ajaygit158-zz** [comment]
Jenkins Build is failing. Closing and reopening the PR.

**Comment by ajaygit158-zz** [comment]
Jenkins Build was failing. Closing and reopening the PR.

**Comment by ajaygit158-zz** [comment]
@vrozov Kindly review

**Comment by ajaygit158-zz** [comment]
@tushargosavi Kindly review the changes

File: engine/src/main/java/com/datatorrent/stram/client/DTConfiguration.java (hunk 1 above)
Comment (sandeshh) [reply engine/src/main/java/com/datatorrent/stram/client/DTConfiguration.java]: DTConfiguration exposes iterators, that should be used. Map is an internal representation which should not be exposed.

File: engine/src/main/java/com/datatorrent/stram/client/DTConfiguration.java [hunk 4]
```diff
@@ -73,7 +74,9 @@
   public static class ValueEntry
```
Comment (sandeshh) [inline engine/src/main/java/com/datatorrent/stram/client/DTConfiguration.java]: Please make this private. Introduce PropertyInfo class in AppPackage, with value and description.

You can populate the description field from DtConfiguration.getDescription method

**Comment by ajaygit158-zz** [comment]
@sandeshh Incorporated all your review comments.

File: engine/src/main/java/com/datatorrent/stram/client/DTConfiguration.java [hunk 5]
```diff
@@ -70,10 +71,12 @@
   private final Map<String, ValueEntry> map = new LinkedHashMap<>();
   private static final Logger LOG = LoggerFactory.getLogger(DTConfiguration.class);
 
-  public static class ValueEntry
+  private static class ValueEntry
   {
     public String value;
+    @JsonIgnore
```
Comment (vrozov) [inline engine/src/main/java/com/datatorrent/stram/client/DTConfiguration.java]: Is it still necessary?

File: engine/src/main/java/com/datatorrent/stram/client/AppPackage.java [hunk 6]
```diff
@@ -426,7 +452,10 @@ private void processPropertiesXml(File file, AppInfo app)
       config.loadFile(file);
       for (Map.Entry<String, String> entry : config) {
         String key = entry.getKey();
+        PropertyInfo propertyInfo = new PropertyInfo();
```
Comment (vrozov) [inline engine/src/main/java/com/datatorrent/stram/client/AppPackage.java]: It is not necessary to construct propertyInfo when `value == null`, move to line 466.

File: engine/src/main/java/com/datatorrent/stram/client/AppPackage.java [hunk 7]
```diff
@@ -109,6 +109,32 @@ public AppInfo(String name, String file, String type)
 
   }
 
+  public static class PropertyInfo
+  {
+    private String value;
```
Comment (vrozov) [inline engine/src/main/java/com/datatorrent/stram/client/AppPackage.java]: Both value and description should be final.

**Review by vrozov — CHANGES_REQUESTED** [review]
squash commits.

File: engine/src/main/java/com/datatorrent/stram/cli/ApexCli.java [hunk 8]
```diff
@@ -3474,13 +3475,25 @@ public void execute(String[] args, ConsoleReader reader) throws Exception
...
+        for (Map.Entry<String, PropertyInfo> entry : ap.getDefaultProperties().entrySet()) {
+          if (entry.getKey().contains(".attr.")) {
```
Comment (sandeshh) [inline engine/src/main/java/com/datatorrent/stram/cli/ApexCli.java]: While creating the PropertyInfo don't load the Description for attributes. So you don't need both the for loops.

File: engine/src/main/java/com/datatorrent/stram/client/AppPackage.java [hunk 9]
```diff
@@ -435,10 +464,10 @@ private void processPropertiesXml(File file, AppInfo app)
           }
         } else {
           if (app == null) {
-            defaultProperties.put(key, value);
+            defaultProperties.put(key, propertyInfo);
```
Comment (sandeshh) [inline engine/src/main/java/com/datatorrent/stram/client/AppPackage.java]: While creating Property make description null for attributes

File: engine/src/main/java/com/datatorrent/stram/client/DTConfiguration.java (hunk 5 above)
Comment (ajaygit158-zz) [reply engine/src/main/java/com/datatorrent/stram/client/DTConfiguration.java]: Forgot to remove. Removed this in updated PR.

File: engine/src/main/java/com/datatorrent/stram/client/AppPackage.java (hunk 7 above)
Comment (ajaygit158-zz) [reply engine/src/main/java/com/datatorrent/stram/client/AppPackage.java]: Changed

File: engine/src/main/java/com/datatorrent/stram/client/AppPackage.java (hunk 6 above)
Comment (ajaygit158-zz) [reply engine/src/main/java/com/datatorrent/stram/client/AppPackage.java]: changed

File: engine/src/main/java/com/datatorrent/stram/cli/ApexCli.java (hunk 8 above)
Comment (ajaygit158-zz) [reply engine/src/main/java/com/datatorrent/stram/cli/ApexCli.java]: Updated the code for this change.

**Review by sandeshh — APPROVED** [review]
Please squash the commits. If @vrozov is fine will merge it.

**Comment by ajaygit158-zz** [comment]
Build failed for non related test. Closing and Reopening PR

**Comment by ajaygit158-zz** [comment]
Squashed the commits.

**Review by vrozov — APPROVED** [review]
@sandeshh Please merge.

**Pull request merged**